Use a `WeakMap` in `src/display/text_layer.js`

timvandermeij opened the pull request:
This is a part of #7584 that fixes the third and fourth item. Look at the commit messages for more details.

timvandermeij committed:
Refactor `text_layer.js` to pass the task as a parameter

We pass many parameters to `appendText` while we might as well pass the
`task` object that contains them. This saves a few lines of code and
makes the signature of `appendText` more clear. We do the same for
`expand`, which is useful for the next commit in which we replace
`div.dataset` with a `WeakMap`.

Furthermore, this patch adds a missing parameter to a comment block to
make it clear which parameters remain.

timvandermeij committed:
Replace `div.dataset` with a `WeakMap` in `text_layer.js`

This patch improves performance by avoiding unnecessary type
conversions, which also help the JIT for optimizations.

Moreover, this patch fixes issues with the div expansion code where
`textScale` would be undefined in a division. Because of the `dataset`
usage, other comparisons evaluated to `true` while `false` would have
been correct. This makes the expansion mode now work correctly for cases
with, for example, each glyph in one div.

The polyfill for `WeakMap` has been provided by @yurydelendik.

timvandermeij commented:
/botio-linux preview

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -46,6 +46,8 @@ var getDefaultSetting = displayDOMUtils.getDefaultSetting;
  *   initially be set to empty array.
  * @property {number} timeout - (optional) Delay in milliseconds before
  *   rendering of the text  runs occurs.
+ * @property {boolean} enhanceTextSelection - Whether to turn on the text
```
Nit: similar to the `timeout` property above, let's add "(optional)" to this doc string.

timvandermeij replied:
Fixed in the new commit.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -585,7 +594,7 @@ var renderTextLayer = (function renderTextLayerClosure() {
           div = this._textDivs[i];
           div.style.padding = 0;
-          transform = div.dataset.originalTransform || '';
+          transform = divProperties.originalTransform || '';
```
`divProperties` isn't defined in the `else` branch, which probably explains why the text layer isn't reset properly on mouseup.

timvandermeij replied:
Fixed in the new commit. The structure was a bit strange here since the `for` loop and `div` fetching is equal in both the `if` and `else` branches, so I've slightly refactored that to prevent this issue.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

timvandermeij commented:
I updated the second patch to fix the debugger as it relied on `dataset` for the font name. Since the font name is only required for the debugger and not for anything else, it's fine to use `dataset` only for that field (and one might argue that it does not even belong in `textDivProperties` if it's only for the debugger). Manual testing confirms that the debugger now works again. For completeness, this is the diff with the tested version above, which should have no effect on the test run:

```diff
diff --git a/src/display/text_layer.js b/src/display/text_layer.js
index 3d57c58..86a85b4 100644
--- a/src/display/text_layer.js
+++ b/src/display/text_layer.js
@@ -65,7 +65,6 @@ var renderTextLayer = (function renderTextLayerClosure() {
     var textDivProperties = {
       angle: 0,
       canvasWidth: 0,
-      fontName: '',
       isWhitespace: false,
       originalTransform: '',
       originalWidth: 0,
@@ -113,9 +112,10 @@ var renderTextLayer = (function renderTextLayerClosure() {
     textDiv.textContent = geom.str;
     // |fontName| is only used by the Font Inspector. This test will succeed
     // when e.g. the Font Inspector is off but the Stepper is on, but it's
-    // not worth the effort to do a more accurate test.
+    // not worth the effort to do a more accurate test. We only use `dataset`
+    // here to make the font name available for the debugger.
     if (getDefaultSetting('pdfBug')) {
-      textDivProperties.fontName = geom.fontName;
+      textDiv.dataset.fontName = geom.fontName;
     }
     if (angle !== 0) {
       textDivProperties.angle = angle * (180 / Math.PI);
```

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -513,15 +516,8 @@ var renderTextLayer = (function renderTextLayerClosure() {
       for (var i = 0, len = textItems.length; i < len; i++) {
-        appendText(textDivs, viewport, textItems[i], styles, bounds,
-                   enhanceTextSelection);
+        appendText(this, textItems[i]);
```
Since you're passing in `textItems` separately here, I'm wondering if it wouldn't make sense to still pass in `styles` in a similar way too?
E.g. re-add `var styles = this._textContent.styles;` above and make this function call `appendText(this, textItems[i], styles);`?

The reason that I'm asking is that this would get rid of the, comparatively complex given the others, assignment `var style = task._textContent.styles[geom.fontName];` in `appendText` and use the previous `var style = styles[geom.fontName];` instead.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -540,55 +561,53 @@ var renderTextLayer = (function renderTextLayerClosure() {
+
+        if (expandDivs) {
+          var transform = '';
+          var width = divProperties.originalWidth;
```
Nit: Since you're directly accessing `divProperties.canvasWidth` below, I think that you can probably do the same thing with `divProperties.originalWidth` as well and not use the temporary `width` variable.

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -540,55 +561,53 @@ var renderTextLayer = (function renderTextLayerClosure() {
           }
-          var rotation = div.dataset.angle;
-          if (rotation) {
+          var rotation = divProperties.angle;
```
Nit: Again, you might just get rid of this variable and use `divProperties.angle` directly below (similar to the `divProperties.padding...` ones below).

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -540,59 +557,57 @@ var renderTextLayer = (function renderTextLayerClosure() {
+        } else {
           div.style.padding = 0;
-          transform = div.dataset.originalTransform || '';
+          transform = divProperties.originalTransform;
```
Isn't `transform` only defined in the `else` block thanks to [variable hoisting](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var#var_hoisting), which is something that we generally try to avoid. You could probably just do `CustomStyle.setProp('transform', div, divProperties.originalTransform);` instead.

timvandermeij commented:
The review comments have been addressed. These are the difference since the last version:

```diff
diff --git a/src/display/text_layer.js b/src/display/text_layer.js
index 86a85b4..f1aa9b8 100644
--- a/src/display/text_layer.js
+++ b/src/display/text_layer.js
@@ -59,7 +59,7 @@ var renderTextLayer = (function renderTextLayerClosure() {
     return !NonWhitespaceRegexp.test(str);
   }
 
-  function appendText(task, geom) {
+  function appendText(task, geom, styles) {
     // Initialize all used properties to keep the caches monomorphic.
     var textDiv = document.createElement('div');
     var textDivProperties = {
@@ -83,7 +83,7 @@ var renderTextLayer = (function renderTextLayerClosure() {
 
     var tx = Util.transform(task._viewport.transform, geom.transform);
     var angle = Math.atan2(tx[1], tx[0]);
-    var style = task._textContent.styles[geom.fontName];
+    var style = styles[geom.fontName];
     if (style.vertical) {
       angle += Math.PI / 2;
     }
@@ -210,8 +210,8 @@ var renderTextLayer = (function renderTextLayerClosure() {
 
       var transform = '';
       if (textDivProperties.canvasWidth !== 0 && width > 0) {
-        var textScale = textDivProperties.canvasWidth / width;
-        transform = 'scaleX(' + textScale + ')';
+        var scale = textDivProperties.canvasWidth / width;
+        transform = 'scaleX(' + scale + ')';
       }
       var rotation = textDivProperties.angle;
       if (rotation !== 0) {
@@ -537,8 +537,9 @@ var renderTextLayer = (function renderTextLayerClosure() {
 
     _render: function TextLayer_render(timeout) {
       var textItems = this._textContent.items;
+      var textStyles = this._textContent.styles;
       for (var i = 0, len = textItems.length; i < len; i++) {
-        appendText(this, textItems[i]);
+        appendText(this, textItems[i], textStyles);
       }
 
       if (!timeout) { // Render right away
@@ -568,22 +569,21 @@ var renderTextLayer = (function renderTextLayerClosure() {
 
         if (expandDivs) {
           var transform = '';
-          var width = divProperties.originalWidth;
-          var textScale = 1;
+          var scale = 1;
 
-          if (divProperties.canvasWidth !== 0 && width > 0) {
-            textScale = divProperties.canvasWidth / width;
-            transform = 'scaleX(' + textScale + ')';
+          if (divProperties.canvasWidth !== 0 &&
+              divProperties.originalWidth > 0) {
+            scale = divProperties.canvasWidth / divProperties.originalWidth;
+            transform = 'scaleX(' + scale + ')';
           }
-          var rotation = divProperties.angle;
-          if (rotation !== 0) {
-            transform = 'rotate(' + rotation + 'deg) ' + transform;
+          if (divProperties.angle !== 0) {
+            transform = 'rotate(' + divProperties.angle + 'deg) ' + transform;
           }
           if (divProperties.paddingLeft !== 0) {
             div.style.paddingLeft =
-              (divProperties.paddingLeft / textScale) + 'px';
+              (divProperties.paddingLeft / scale) + 'px';
             transform += ' translateX(' +
-              (-divProperties.paddingLeft / textScale) + 'px)';
+              (-divProperties.paddingLeft / scale) + 'px)';
           }
           if (divProperties.paddingTop !== 0) {
             div.style.paddingTop = divProperties.paddingTop + 'px';
@@ -591,7 +591,7 @@ var renderTextLayer = (function renderTextLayerClosure() {
           }
           if (divProperties.paddingRight !== 0) {
             div.style.paddingRight =
-              divProperties.paddingRight / textScale + 'px';
+              divProperties.paddingRight / scale + 'px';
           }
           if (divProperties.paddingBottom !== 0) {
             div.style.paddingBottom = divProperties.paddingBottom + 'px';
@@ -601,8 +601,8 @@ var renderTextLayer = (function renderTextLayerClosure() {
           }
         } else {
           div.style.padding = 0;
-          transform = divProperties.originalTransform;
-          CustomStyle.setProp('transform', div, transform);
+          CustomStyle.setProp('transform', div,
+                              divProperties.originalTransform);
         }
       }
     },
```

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -185,39 +205,42 @@ var renderTextLayer = (function renderTextLayerClosure() {
-        transform = '';
+      var transform = '';
+      if (textDivProperties.canvasWidth !== 0 && width > 0) {
+        var scale = textDivProperties.canvasWidth / width;
```
I'm *really* sorry that I didn't think to ask about this before now, but I wonder if you couldn't store the `scale` in `textDivProperties.scale` (with `1` as the default value [above](https://github.com/mozilla/pdf.js/pull/7588/files#diff-986df9c532f6b9ab6cf81a23ce26e446R65)) instead?

That way you'd not need to either store `textDivProperties.originalWidth`,  nor recompute the `scale` in `expandTextDivs`. Instead, [below](https://github.com/mozilla/pdf.js/pull/7588/files#diff-986df9c532f6b9ab6cf81a23ce26e446R572) you'd just do e.g.
```javascript
...
var scale = divProperties.scale;

if (scale !== 1) {
  transform = 'scaleX(' + scale + ')';
}
...
```

timvandermeij replied:
Excellent idea! I have done this in a separate commit (as it's not really related to the `WeakMap` or refactoring, but more to optimization).

Snuffleupagus commented:
This looks great, thank you for working on this!

After looking through this once more, I've got one remaining question.
`r=me`, with or without https://github.com/mozilla/pdf.js/pull/7588#discussion_r77453205 addressed, so just go ahead and land this as-is if you prefer!

(We can improve this further in a follow-up, since I don't want unnecessarily delay landing this more just because of a last minute idea on my part.)

timvandermeij committed:
Optimize scale calculation in `text_layer.js`

This patch avoids having to calculate the scale twice by saving it in
the properties object.

Moreover, we remove a temporary variable and place parentheses around a
calculation inside a string concatenation.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

Snuffleupagus commented:
Thank you for the patch!

/botio makeref
